chore: use the bb predicates when generating constraints

guipublic opened the pull request:
This PR adds support for using the acir predicates for some black-box function calls on barretenberg side
- For recursive proving, we conditionally assign the inputs when the predicate is false to a placeholder verifiable proof and vk
- For elliptic curve points, we conditionally assign the inputs when the predicate is false, to valid dummy values

ecdsa r1 is not done on purpose (it'll be merged into one ecdsa file)

iakovenkos commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/witness_constant.cpp:
```diff
@@ -40,6 +40,31 @@ bb::stdlib::cycle_group<Builder> to_grumpkin_point(const WitnessOrConstant<FF>&
+    }
+    // creates a bool_ct from the index, without adding a boolean gate because a predicate is boolean by definition.
+    FF predicate_value = builder.get_variable(predicate.index);
+    bool_ct predicate_witness = bool_ct(&builder, predicate_value == FF(1));
```
I don't have enough context here, but it seems a bit suspicious that any `predicate_value` except for `1` is getting converted to `0`. Can you please explain why it's safe?

iakovenkos replied:
Is it because inside of the `to_grumkin_point(...)` we have
`auto infinite = bool_ct(to_field_ct(input_infinite, builder));`?
Should we maybe add a `from_bool_witness_index` in `bool_t`?

guipublic replied:
The predicate is coming from Noir, as a condition guarding a conditional branch.
For instance, if you have:
if x ==3 { C = ec_add(A,B); }, the predicate will be x==3, which is a boolean value. When it is computed by Noir, it is properly constrained to be boolean.

>  Should we maybe add a from_bool_witness_index in bool_t?

Yes that could be useful for this case.

iakovenkos replied:
Thanks for the explanation! I'm spinning up a PR to add this constructor

iakovenkos replied:
my PR got merged to next, so you can use `from_witness_undex_unsafe()` on `bool_t`

TomAFrench commented:
@guipublic looks like this has been affected by changes the crypto team has made to how we handle ECDSA opcodes. Can you update this please?

ledwards2225 commented:
@guipublic hope this hasn't been waiting on me - I got the sense that it wasn't ready for review yet. Lmk!

TomAFrench commented:
Nope, don't worry. This has been semi-blocked by us getting the acir serialization PR in until recently.

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/ecdsa_constraints.cpp:
```diff
@@ -91,6 +91,12 @@ void create_ecdsa_verify_constraints(typename Curve::Builder& builder,
     Fq pub_x(pub_x_bytes);
     Fq pub_y(pub_y_bytes);
     G1 public_key(pub_x, pub_y);
+    // Update it depending on the predicate
```
Would be good to have @federicobarbacovi take a quick look since this code has been audited

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/ecdsa_constraints.cpp:
```diff
@@ -91,6 +91,12 @@ void create_ecdsa_verify_constraints(typename Curve::Builder& builder,
     G1 public_key(pub_x, pub_y);
+    // Update it depending on the predicate
+    if (!input.predicate.is_constant) {
+        bool_ct predicate_witness = bool_ct::from_witness_index_unsafe(&builder, input.predicate.index, true);
```
```suggestion
        bool_ct predicate_witness = bool_ct::from_witness_index_unsafe(&builder, input.predicate.index, /*witness_inverted=*/true);
```

ledwards2225 replied:
Its not clear to me why we need to set witness_inverted = true here. Can we add some explanation?

guipublic replied:
Done

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/honk_recursion_constraint.cpp:
```diff
@@ -129,15 +131,77 @@ HonkRecursionConstraintOutput<typename Flavor::CircuitBuilder> create_honk_recur
         ProofSurgeon<uint256_t>::create_indices_for_reconstructed_proof(input.proof, input.public_inputs);
     stdlib::Proof<Builder> proof_fields = RecursionConstraint::fields_from_witnesses(builder, proof_indices);
 
+    // Reucrsion constraints come with a predicate (e.g. when the black-box call is done in an if conditional depending
```
```suggestion
    // Recursion constraints come with a predicate (e.g. when the black-box call is done in an if conditional depending
``

guipublic replied:
Done

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/honk_recursion_constraint.cpp:
```diff
@@ -129,15 +131,77 @@ HonkRecursionConstraintOutput<typename Flavor::CircuitBuilder> create_honk_recur
+        vk_hash = field_ct<Builder>::conditional_assign(
+            predicate_witness, vk_hash, field_ct<Builder>::from_witness(&builder, mock_vk_hash.get_value()));
+        proof_fields = result;
+        key_fields = result_vk;
```
I think the above logic could use some work. Here are some suggestions:
- Abstract away all of this complex logic into a standalone method designed to return the correct proof/vk/hash given the situation
- Ensure consistent naming: e.g. why "place_holder" for the proof/vk but "mock" for the vk_hash? "result" for proof and "result_vk" for vk?
- Add some explanation for why we use different mock data depending on has_valid_witness_assignments (again, a good place to do this is in the @details of the standalone method)
- Consistent use of from_witness vs construction from witness_ct. (Convention says prefer the former)

guipublic replied:
Done

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/honk_recursion_constraint.test.cpp:
```diff
@@ -137,7 +138,9 @@ template <typename RecursiveFlavor> class AcirHonkRecursionConstraint : public :
     template <typename BuilderType>
-    BuilderType create_outer_circuit(std::vector<InnerBuilder>& inner_circuits, bool dummy_witnesses = false)
+    BuilderType create_outer_circuit(std::vector<InnerBuilder>& inner_circuits,
+                                     bool dummy_witnesses = false,
```
I tend to think we should avoid defualted parameters in almost all cases but certainly in tests. It seems much better to be explicit about what we're testing

guipublic replied:
Done

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/mock_verifier_inputs.cpp:
```diff
@@ -180,6 +180,66 @@ template <typename Flavor, class PublicInputs> HonkProof create_mock_honk_proof(
+    }
+
+    // prove the circuit constructed above
+    // Create the decider proving key
```
Comments like "construct foo" should be avoided since the code itself is documentation

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/mock_verifier_inputs.cpp:
```diff
@@ -180,6 +180,66 @@ template <typename Flavor, class PublicInputs> HonkProof create_mock_honk_proof(
+    // constuct a circuit with a single gate
+    Builder builder;
+
+    // Create 2^log_n many add gates based on input log num gates
```
delete?

ledwards2225 replied:
flagging this again

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/multi_scalar_mul.test.cpp:
```diff
@@ -50,12 +50,22 @@ TEST_F(MSMTests, TestMSM)
+    // Create a predicate witness or constant which takes the index of the last witness in the array
+    auto predicate = WitnessOrConstant<fr>::from_index(static_cast<uint32_t>(witness.size()));
+    witness.push_back(fr(0));
+    // Set the predicate in the msm_constrain
```
Another comment that takes up space without adding any value

guipublic replied:
Done

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/multi_scalar_mul.test.cpp:
```diff
@@ -50,12 +50,22 @@ TEST_F(MSMTests, TestMSM)
+    };
+    // Create a predicate witness or constant which takes the index of the last witness in the array
+    auto predicate = WitnessOrConstant<fr>::from_index(static_cast<uint32_t>(witness.size()));
+    witness.push_back(fr(0));
```
Isn't this effectively nullifying the single test we had of this MSM logic but setting the predicate to false? At the very least we should have a nominal test and a test that shows invalid constraints are ignored by a false predicate

guipublic replied:
I added several tests to cover most cases.

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/witness_constant.cpp:
```diff
@@ -40,6 +40,29 @@ bb::stdlib::cycle_group<Builder> to_grumpkin_point(const WitnessOrConstant<FF>&
+    // creates a bool_ct from the index, without adding a boolean gate because a predicate is boolean by definition.
+    bool_ct predicate_witness = bool_ct::from_witness_index_unsafe(&builder, predicate.index);
+
+    auto g1 = bb::grumpkin::g1::affine_one;
```
A bit of a nit but I think g1 is a confusing name here because its conflating the "1" from the identifier of the group G1 with the generator "one" in that group. I think the name `one` is better or `generator` if you like

guipublic replied:
Done

ledwards2225 commented on barretenberg/cpp/src/barretenberg/stdlib/primitives/bool/bool.cpp:
```diff
@@ -93,7 +93,9 @@ bool_t<Builder>::bool_t(bool_t<Builder>&& other)
-bool_t<Builder> bool_t<Builder>::from_witness_index_unsafe(Builder* ctx, const uint32_t witness_index)
+bool_t<Builder> bool_t<Builder>::from_witness_index_unsafe(Builder* ctx,
+                                                           const uint32_t witness_index,
+                                                           const bool witness_inverted)
```
We should run this by @iakovenkos if we haven't already, especially because bool is now out for audit

guipublic replied:
Done,
as discussed with @iakovenkos, I have removed the 'witness_inverted' default parameter that is not needed anymore.

ledwards2225 commented on barretenberg/cpp/src/barretenberg/stdlib/primitives/databus/databus.test.cpp:
```diff
@@ -117,7 +117,6 @@ TEST(Databus, UnnormalizedEntryAccess)
     std::vector<field_ct> calldata_entries;
     for (fr entry : raw_calldata_entries) {
         calldata_entries.emplace_back(witness_ct(&builder, entry));
```
unrelated change?

guipublic replied:
This line is giving me compilation error because `entry_witness` is not used.

ledwards2225 reviewed (changes requested):
Overall these changes seem ok to me but I think there's still some work to do to get this to a merge-able state. My main concerns are 
- Ensuring all of this new functionality is fully tested
- Making sure @iakovenkos and @federicobarbacovi are happy with and aware of the changes in bool / ecdsa respectively
- Making sure the complex portions (especially of verifier input mocking) are isolated, clean and well documented

federicobarbacovi commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/ecdsa_constraints.cpp:
```diff
@@ -91,6 +91,16 @@ void create_ecdsa_verify_constraints(typename Curve::Builder& builder,
     Fq pub_y(pub_y_bytes);
     G1 public_key(pub_x, pub_y);
+    // Update it depending on the predicate
+    if (!input.predicate.is_constant) {
```
```suggestion
    if (!input.predicate.is_constant) {
        // We need to use a point which is different from the generator otherwise secp256r1 ECDSA verification fails
        auto default_point = Curve::g1::one + Curve::g1::one;
        bool_ct predicate_witness =
            bool_ct::from_witness_index_unsafe(&builder, input.predicate.index);
        pub_x = Fq::conditional_assign(predicate_witness, pub_x, default_point.x);
        pub_y = Fq::conditional_assign(predicate_witness, pub_y, default_point.y);
    } else {
        BB_ASSERT_EQ(input.predicate.value, true, "Creating ECDSA constraints with a constant predicate equal to false.");
    }
```

federicobarbacovi replied:
`conditional_assign` works in reverse respect to `conditional_select`, I think the code as written above might be easier to read. Also, we need to use a point which is not ± the generator to avoid issues with secp256r1

Could you please also add documentation regarding this change in the doxygen documentation for the function `create_ecdsa_verify_constraints`, and documentation regarding `predicate` to the doxygen documentation of the class `EcdsaConstraint`? Thanks!

Finally, I think it would be good to add some tests in `ecdsa_constraints.test.cpp` testing that when we supply the predicate things work as expected.

guipublic replied:
Done

TomAFrench commented on barretenberg/acir_tests/bootstrap.sh:
```diff
@@ -93,7 +93,7 @@ function regenerate_recursive_inputs {
   mv ./target/assert_statement.gz ./target/witness.gz
   cd ../..
-  parallel 'run_proof_generation {}' ::: $(ls internal_test_programs)
+  parallel 'run_proof_generation {}' ::: "double_verify_honk_proof" "verify_honk_proof" "verify_rollup_honk_proof"
```
```suggestion
  parallel 'run_proof_generation {}' ::: "double_verify_honk_proof" "verify_honk_proof" "verify_honk_zk_proof" "verify_rollup_honk_proof"
```

TomAFrench commented on barretenberg/acir_tests/bootstrap.sh:
```diff
@@ -93,7 +93,7 @@ function regenerate_recursive_inputs {
   mv ./target/assert_statement.gz ./target/witness.gz
   cd ../..
-  parallel 'run_proof_generation {}' ::: $(ls internal_test_programs)
+  parallel 'run_proof_generation {}' ::: "double_verify_honk_proof" "verify_honk_proof" "verify_honk_zk_proof" "verify_rollup_honk_proof"
```
```suggestion
  parallel 'run_proof_generation {}' ::: "double_verify_honk_proof" "verify_honk_proof" "verify_honk_zk_proof" "double_verify_honk_zk_proof" "verify_rollup_honk_proof"
```

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/ecdsa_constraints.test.cpp:
```diff
@@ -173,6 +173,66 @@ TYPED_TEST(EcdsaConstraintsTest, GenerateVKFromConstraints)
+        AcirProgram program{ constraint_system, witness_values };
+        auto builder = create_circuit<Builder>(program);
+
+        info("Num gates: ", builder.get_estimated_num_finalized_gates());
```
Do we want to keep these prints? If the gate counts are interesting one option is to actually "pin" them in the test by asserting the value. This has the nice benefit of making them statically readable

guipublic replied:
IMHO it's fine to have this inside a test, but I'm fine to remove them as well, not sure for the assert.
@federicobarbacovi do you want to keep them?

guipublic replied:
keeping them for now

ledwards2225 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/honk_recursion_constraint.cpp:
```diff
@@ -89,6 +89,64 @@ void create_dummy_vkey_and_proof(typename Flavor::CircuitBuilder& builder,
+    } else {
+        // If we have an actual witness, the place holder proof and VK should be an actual verifiable honk proof and
+        // honk vk. Otherwise the recursive verification would fail if the predicate is false.
+        auto [proof, place_holder_vk] =
```
nit: can we name proof and vk consistently here? Or is there some reason why only the vk is labeled placeholder?

guipublic replied:
Done, proof is an intermediate variable, but in fact it is not needed. I removed it.

ledwards2225 reviewed (approved):
LGTM, thanks for making updates!

TomAFrench commented on .test_patterns.yml:
```diff
@@ -313,6 +313,11 @@ tests:
+    error_regex: "Failed to add 48 validators. Active validators: 0 -> 47. Queue: 0 -> 0. A likely issue is the bootstrap size."
+    owners:
+      - *lasse
+
```
```suggestion
```

Should be fixed by https://github.com/AztecProtocol/aztec-packages/pull/17594

TomAFrench reviewed (approved):
LGTM

guipublic committed:
chore: use the bb predicates when generating constraints

This PR adds support for using the acir predicates for some black-box function calls on barretenberg side
- For recursive proving, we conditionally assign the inputs when the predicate is false to a placeholder verifiable proof and vk
- For elliptic curve points, we conditionally assign the inputs when the predicate is false, to valid dummy values

ecdsa r1 is not done on purpose (it'll be merged into one ecdsa file)

Co-authored-by: Alejo Amiras <alejo.amiras@gmail.com>
Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: Ary Borenszweig <asterite@gmail.com>
Co-authored-by: Charlie <5764343+charlielye@users.noreply.github.com>
Co-authored-by: Charlie Lye <5764343+charlielye@users.noreply.github.com>
Co-authored-by: DanielKotov <159419107+DanielKotov@users.noreply.github.com>
Co-authored-by: David Banks <47112877+dbanks12@users.noreply.github.com>
Co-authored-by: Esau <esau@aztecprotocol.com>
Co-authored-by: Facundo <fcarreiro@users.noreply.github.com>
Co-authored-by: Gregorio Juliana <gregojquiros@gmail.com>
Co-authored-by: Harsh Bajpai <bajpaiharsh244@gmail.com>
Co-authored-by: Ilyas Ridhuan <ilyas@aztecprotocol.com>
Co-authored-by: IlyasRidhuan <ilyasridhuan@gmail.com>
Co-authored-by: Innokentii Sennovskii <isennovskiy@gmail.com>
Co-authored-by: James Zaki <james.zaki@proton.me>
Co-authored-by: Jan Beneš <janbenes1234@gmail.com>
Co-authored-by: Jean M <132435771+jeanmon@users.noreply.github.com>
Co-authored-by: Jonathan Hao <johnathan79717@gmail.com>
Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>
Co-authored-by: Jonathan Hao <jonathanpohsianghao@gmail.com>
Co-authored-by: Josh Crites <jc@joshcrites.com>
Co-authored-by: José Pedro Sousa <jose@aztecprotocol.com>
Co-authored-by: José Pedro Sousa <outgoing@zpedro.dev>
Co-authored-by: Khashayar Barooti <khashayar@aztecprotocol.com>
Co-authored-by: LHerskind <16536249+LHerskind@users.noreply.github.com>
Co-authored-by: Lasse Herskind <16536249+LHerskind@users.noreply.github.com>
Co-authored-by: Leila Wang <leizciw@gmail.com>
Co-authored-by: Lin Oshitani <linoshitani@gmail.com>
Co-authored-by: Lucas Xia <lucasxia01@gmail.com>
Co-authored-by: Maddiaa <47148561+Maddiaa0@users.noreply.github.com>
Co-authored-by: Maddiaa0 <47148561+Maddiaa0@users.noreply.github.com>
Co-authored-by: Michael Connor <mike@aztecprotocol.com>
Co-authored-by: Miranda Wood <miranda@aztecprotocol.com>
Co-authored-by: MirandaWood <miranda@aztecprotocol.com>
Co-authored-by: Mitch <mitchell@aztecprotocol.com>
Co-authored-by: Mitchell Tracy <mitchellftracy@gmail.com>
Co-authored-by: Nicolás Venturo <nicolas.venturo@gmail.com>
Co-authored-by: Paperclip Minimizer <minim@wonderland.xyz>
Co-authored-by: PhilWindle <60546371+PhilWindle@users.noreply.github.com>
Co-authored-by: PhilWindle <philip.windle@gmail.com>
Co-authored-by: Ragnar <rodiondenmark@gmail.com>
Co-authored-by: Rahul Kothari <rahul.kothari.201@gmail.com>
Co-authored-by: Raju Krishnamoorthy <krishnamoorthy@gmail.com>
Co-authored-by: Rumata888 <isennovskiy@gmail.com>
Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>
Co-authored-by: Santiago Palladino <santiago@aztecprotocol.com>
Co-authored-by: Sarkoxed <75146596+Sarkoxed@users.noreply.github.com>
Co-authored-by: Sarkoxed <sarkoxed2013@yandex.ru>
Co-authored-by: Savio <72797635+Savio-Sou@users.noreply.github.com>
Co-authored-by: StoneMac65 <StoneMac65@gmail.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: Tom French <15848336+TomAFrench@users.noreply.github.com>
Co-authored-by: Zachary James Williamson <blorktronics@gmail.com>
Co-authored-by: benesjan <13470840+benesjan@users.noreply.github.com>
Co-authored-by: benesjan <janbenes1234@gmail.com>
Co-authored-by: critesjosh <18372439+critesjosh@users.noreply.github.com>
Co-authored-by: danielntmd <162406516+danielntmd@users.noreply.github.com>
Co-authored-by: danielntmd <danielntmd@nethermind.io>
Co-authored-by: dbanks12 <david@aztec-labs.com>
Co-authored-by: defkit <84741533+defkit@users.noreply.github.com>
Co-authored-by: esau <152162806+sklppy88@users.noreply.github.com>
Co-authored-by: fcarreiro <facundo@aztecprotocol.com>
Co-authored-by: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Co-authored-by: iAmMichaelConnor <mike@aztecprotocol.com>
Co-authored-by: jeanmon <jean@aztec-labs.com>
Co-authored-by: jfecher <jfecher11@gmail.com>
Co-authored-by: josh crites <critesjosh@gmail.com>
Co-authored-by: josh crites <jc@joshcrites.com>
Co-authored-by: just-mitch <68168980+just-mitch@users.noreply.github.com>
Co-authored-by: ledwards2225 <98505400+ledwards2225@users.noreply.github.com>
Co-authored-by: ledwards2225 <l.edwards.d@gmail.com>
Co-authored-by: lucasxia01 <lucasxia01@gmail.com>
Co-authored-by: ludamad <163993+ludamad@users.noreply.github.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: ludamad <domuradical@gmail.com>
Co-authored-by: maramihali <mara@aztec-labs.com>
Co-authored-by: maramihali <mara@aztecprotocol.com>
Co-authored-by: mralj <nikola.mratinic@gmail.com>
Co-authored-by: nventuro <2530770+nventuro@users.noreply.github.com>
Co-authored-by: saleel <saleel@aztecprotocol.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
Co-authored-by: sirasistant <sirasistant@gmail.com>
Co-authored-by: spypsy <spypsy@users.noreply.github.com>
Co-authored-by: thunkar <gregojquiros@gmail.com>
Co-authored-by: Álvaro Rodríguez <sirasistant@gmail.com>
